File: rows/address.py
"""Postal address of a certain location"""


class Address:
    """Postal address of a certain location"""

    ROAD = 'road'
    HOUSE_NUMBER = 'house_number'
    CITY = 'city'
    COUNTRY_CODE = 'country_code'
    COUNTRY = 'country'
    NEIGHBOURHOOD = 'neighbourhood'
    POSTCODE = 'postcode'
    STATE = 'state'
    SUBURB = 'suburb'

    def __init__(self, **kwargs):
        self.__house_number = kwargs.get(Address.HOUSE_NUMBER, None)
        self.__road = kwargs.get(Address.ROAD, None)
        self.__city = kwargs.get(Address.CITY, None)
        self.__country_code = kwargs.get(Address.COUNTRY_CODE, None)
        self.__country = kwargs.get(Address.COUNTRY, None)
        self.__neighbourhood = kwargs.get(Address.NEIGHBOURHOOD, None)
        self.__postcode = kwargs.get(Address.POSTCODE, None)
        self.__state = kwargs.get(Address.STATE, None)
        self.__suburb = kwargs.get(Address.SUBURB, None)

    def __eq__(self, other):
        if not isinstance(other, Address):
            return False

        return self.__house_number == other.house_number and self.__road == other.road and self.__city == other.city \
            and self.__country_code == other.country_code and self.__country == other.country \
            and self.__neighbourhood == other.neighbourhood and self.__postcode == other.postcode \
            and self.__state == other.state and self.__suburb == other.suburb

    def __hash__(self):
        return hash(self.tuple())

    def tuple(self):
        """Returns object as tuple"""

        return (self.__house_number,
                self.__road,
                self.__city,
                self.__country,
                self.__country_code,
                self.__neighbourhood,
                self.__postcode,
                self.__state,
                self.__suburb)

    @property
    def house_number(self):
        """Get property"""

        return self.__house_number

    @property
    def road(self):
        """Get property"""

        return self.__road

    @property
    def city(self):
        """Get property"""

        return self.__city

    @property
    def country(self):
        """Get property"""

        return self.__country

    @property
    def country_code(self):
        """Get property"""

        return self.__country_code

    @property
    def neighbourhood(self):
        """Get property"""

        return self.__neighbourhood

    @property
    def postcode(self):
        """Get property"""

        return self.__postcode

    @property
    def state(self):
        """Get property"""

        return self.__state

    @property
    def suburb(self):
        """Get property"""

        return self.__suburb

File: src/main/address.cpp
#include "address.h"

namespace rows {

    Address::Address()
            : Address("", "", "", "") {}

    Address::Address(std::string house_number, std::string street, std::string city, std::string post_code)
            : house_number_(std::move(house_number)),
              street_(std::move(street)),
              city_(std::move(city)),
              post_code_(std::move(post_code)) {}

    Address::Address(const Address &other)
            : house_number_(other.house_number_),
              street_(other.street_),
              post_code_(other.post_code_),
              city_(other.city_) {}

    Address::Address(Address &&other) noexcept {
        house_number_ = std::move(other.house_number_);
        street_ = std::move(other.street_);
        post_code_ = std::move(other.post_code_);
        city_ = std::move(other.city_);
    }

    Address &Address::operator=(const Address &other) {
        house_number_ = other.house_number_;
        street_ = other.street_;
        post_code_ = other.post_code_;
        city_ = other.city_;
        return *this;
    }

    Address &Address::operator=(Address &&other) noexcept {
        house_number_ = std::move(other.house_number_);
        street_ = std::move(other.street_);
        post_code_ = std::move(other.post_code_);
        city_ = std::move(other.city_);
        return *this;
    }

    std::ostream &operator<<(std::ostream &out, const Address &address) {
        out << "(" << address.house_number()
            << ", " << address.street()
            << ", " << address.city()
            << ", " << address.post_code()
            << ")";
        return out;
    }

    bool Address::operator==(const Address &other) const {
        return house_number_ == other.house_number_
               && street_ == other.street_
               && city_ == other.city_
               && post_code_ == other.post_code_;
    }

    bool Address::operator!=(const Address &other) const {
        return !this->operator==(other);
    }

    const std::string &Address::house_number() const {
        return house_number_;
    }

    const std::string &Address::street() const {
        return street_;
    }

    const std::string &Address::city() const {
        return city_;
    }

    const std::string &Address::post_code() const {
        return post_code_;
    }
}
